### PR TITLE
Consistent padding across input containers

### PR DESCRIPTION
**Pull request by rohan-kadkol**
## Changes

Addresses an after-PR TODO: https://github.com/iTwin/iTwinUI/pull/1753#discussion_r1449467506.

The issue was that the visually perceivable space between the end icon and the right border was inconsistent across iui-input-flex-container and iui-input-with-icon.

|Component|Image|Spacing on the end icon|
|-|-|-|
|iui-input-flex-container|image|icon padding left\|right = 16px\|16px|
|iui-input-with-icon|image|icon margin left\|right = 0\|9px|

i.e. icon padding right of iui-input-flex-container (16px) ≠ icon margin right of iui-input-with-icon (9px)

Two options existed:
1. Change the icon margin right of iui-input-with-icon from 9px to 16px
2. Change the icon padding right of iui-input-flex-container from 16px to 9px

#### Option 1
This was not ideal since it would create a lot of extra space to the right of the icon in the Select/ComboBox.

<details>
<summary>Screenshot(s)</summary>

|Old|New|
|-|-|
|image|image|

</details>

A bigger reason against this option was that the new right spacing of 16px will not align with the default padding of 8px on the borderless icon button in the search box.

<details>
<summary>Screenshot(s)</summary>

image|

</details>

#### Option 2
This was the more ideal option since with this, the right icons of iui-input-with-icon and iui-input-flex-container now aligned with the icon button in the search box.

<details>
<summary>Screenshot(s)</summary>

image

</details>

Although the icon button does not look aligned, it is actually aligned. It may not seem so because of some unnecessary padding within the content of the icon:

<details>
<summary>Screenshot(s)</summary>

image

</details>

---

Going with option 2 involved changing the default padding of the icon from being dependent on the size ([code](https://github.com/iTwin/iTwinUI/blob/0b8dc960c60cec4528d845a2af4007d2115ce16a/packages/itwinui-css/src/utils/icon.scss#L29)) (16px in medium icon size) to now be a constant of 8px.

<details>
<summary>Screenshot(s)</summary>

|Old|New|
|-|-|
|image|image|

</details>

This was done so that the 8px inline padding of the icon button now matches with the inline padding for the icon.

Additionally, the inline padding of the icon in iui-input-flex-container earlier did not take into account the 1px border that is positioned absolutely. But now, this is taken into account with a `padding-inline: 1px` on iui-input-flex-container. This prevented the left-most or right-most content from getting overlapped by the border. E.g. textarea's html demo page.

<details>
<summary>Screenshot(s)</summary>

|Old|New|
|-|-|
|Screenshot 2024-02-14 at 5 14 57 PM|Screenshot 2024-02-14 at 5 15 20 PM|

</details>

To keep the padding behavior of `Icon` and `SearchBox.Icon` the same, this PR also removes the custom padding (with magic numbers) on iui-search-icon. Instead, iui-search-icon now uses the `data-iui-padding="true"` to get the same padding given to iui-svg-icon (https://github.com/iTwin/iTwinUI/pull/1828#discussion_r1496690535).

This leads to some misalignment between `SearchBox.Icon` and `SearchBox.Button`, but since keeping the padding behavior consistent between `Icon` and `SearchBox.Icon` is more important, this misalignment is acceptable.

<details>
<summary>Screenshot(s)</summary>

image

</details>

This also leads to a slight layout shift when toggling between the collapsed and expanded states of an expandable SearchBox.

<details>
<summary>Screenshot(s)</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/98cc8497-87f0-4430-9e93-b6d20c348270

</details>

But since most SearchBoxes are positioned on the right, this layout shift is not apparent, and thus is also deemed an acceptable consequence of trying to make the padding behavior across `Icon` and `SearchBox.Icon` consistent.

<details>
<summary>Screenshot(s)</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/3780f973-636a-48b6-a119-f95dcce62ca6

</details>

## Testing

Made the above-mentioned observations by confirmed the failing test images before approving them. Also tested some edge cases in the following modified playgrounds:
* [App.tsx (vite-playground)](https://gist.github.com/rohan-kadkol/f562db2e0aca9392f77695e73fb30430)
* [input.html](https://gist.github.com/rohan-kadkol/5c0ea2c8db3e2a5ee5d7064a60f2d296)

## Docs

Added changesets

### TODOs after PR

* [ ] Fix DOM in the textarea html demos such that the bottom right resizer appears at the bottom right corner of the container and not just the input (screenshot above).
* [ ] Look into making the borderless square button in the `SearchBox` to a regular slightly rectangle button.
  * Or using a regular IconButton without too many customizations.
  * After this, we might have to make some padding adjustments and thus, may also need to modify some changesets.
* [ ] Collapse padding between input and icon in iui-input-flex-container, similar to how it's done in searchbox. (https://github.com/iTwin/iTwinUI/pull/1828#discussion_r1496649623)
  * This should only happen when the buttons are borderless. i.e. the text should not stick to the border of an adjoining item.

### CODE REVIEW

File: packages/itwinui-css/src/searchbox/searchbox.scss [hunk 1]
```diff
@@ -17,6 +17,7 @@
 .iui-expandable-searchbox {
   inline-size: var(--_iui-input-min-height);
   transition: width var(--iui-duration-1) ease-in-out;
+  box-sizing: content-box; // Center the searchbox-open-button (in the non-expanded state) within the expandable-searchbox
```
[the remark this answers is not in the document]
Comment (mayank99) [reply packages/itwinui-css/src/searchbox/searchbox.scss:20]: 🤔 Shouldn't we adjust the padding instead? Changing `box-sizing` feels like a hack

[the remark this answers is not in the document]
Comment (Ben-Pusey-Bentley) [reply packages/itwinui-css/src/searchbox/searchbox.scss:20]: I don't think that using `box-sizing` to center an element is the best solution either.

[the remark this answers is not in the document]
Comment (rohan-kadkol) [reply packages/itwinui-css/src/searchbox/searchbox.scss:20]: Ohh, I see. The reason I was doing this was the fix the below issue in the [SearchBox → Expandable](https://itwin.github.io/iTwinUI/react/?story=searchbox--expandable&theme=dark) when we added `1px` padding-inline on `iui-input-flex-container`:

|Currently hosted version|This PR without `box-sizing: content-box`|This PR with `box-sizing: content-box`|
|-|-|-|
|image|image|image|
|Icon is centered|Icon is right of center|Icon is centered|

The `box-sizing: content-box` was to avoid including the padding into the size. With this, the new size is `38x36` (38 in width since it includes the 36px in content size + 2px in total inline-padding) and thus the icon/button is now centered.

I will look for alternatives to doing this. Open to suggestions.



---

> 🤔 Shouldn't we adjust the padding instead? Changing `box-sizing` feels like a hack

The 1px inline-padding is coming from iui-input-flex-container as a result of this PR. Are you suggesting setting it to 0px when iui-input-flex-container is used in iui-expandable-searchbox?

[the remark this answers is not in the document]
Comment (mayank99) [reply packages/itwinui-css/src/searchbox/searchbox.scss:20]: Yes, we can remove the 1px padding in collapsed state

[the remark this answers is not in the document]
Comment (rohan-kadkol) [reply packages/itwinui-css/src/searchbox/searchbox.scss]: As of now, I removed the 1px padding on .iui-expandable-searchbox regardless of the collapsed/expanded state ([commit](https://github.com/iTwin/iTwinUI/pull/1828/commits/ff19c1c28c4e474aa15d0a14ff5bf30a7c6ffa2b)).

This is because just removing it in the collapsed state is causing an abrupt shift of the search icon when going between the collapsed and expanded states:

https://github.com/iTwin/iTwinUI/assets/45748283/76a6770f-88ba-49c8-bc81-c2e0f07264a1

[the remark this answers is not in the document]
Comment (mayank99) [reply packages/itwinui-css/src/searchbox/searchbox.scss]: Could you maybe use `padding-inline-end` then? It solves the main issue, avoids the layout shift, and simplifies the code.

File: packages/itwinui-css/src/input-container/input-container.scss
Comment (mayank99) [inline packages/itwinui-css/src/input-container/input-container.scss:1]: Do we also need to adjust these magic numbers in `searchbox.scss`?

https://github.com/iTwin/iTwinUI/blob/9896fb2ac2cf793070394610ee319a71e3e60676/packages/itwinui-css/src/searchbox/searchbox.scss#L4-L15

File: packages/itwinui-css/src/searchbox/searchbox.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (rohan-kadkol) [reply packages/itwinui-css/src/searchbox/searchbox.scss]: Yes, just padding-inline-end would solve the main issue and avoid the layout shift (https://github.com/iTwin/iTwinUI/pull/1828/commits/fe605e2cdd6ce2b36d47d6a46a6ace9fb154722a).

But if we want symmetry of the inline-paddings, we can also add a padding-inline-start on all iui-input-flex-containers. Then in the searchbox styles, we can remove padding-inline-start and set padding-inline-end only in the expanded case to prevent the layout shift and fix the searchbox-open-button centering issue. (https://github.com/iTwin/iTwinUI/pull/1828/files/fe605e2cdd6ce2b36d47d6a46a6ace9fb154722a..d90639418b162fc5c6807d2803cf7adaf0eceef7)

[the remark this answers is not in the document]
Comment (mayank99) [reply packages/itwinui-css/src/searchbox/searchbox.scss]: That's fine with me. Can you also add a short comment next to `padding-inline-start: 0`, since it's not obvious why it's there.

[the remark this answers is not in the document]
Comment (rohan-kadkol) [reply packages/itwinui-css/src/searchbox/searchbox.scss]: Added a comment to explain `padding-inline-start: 0` (https://github.com/iTwin/iTwinUI/pull/1828/commits/62f5e5747ea6e4977d341174f6357aa2a69bb18c).

This commit also adds a small comment for `padding-inline: 0`. But I can remove it if its reason is obvious without the comment.

File: packages/itwinui-css/src/input-container/input-container.scss
Comment (rohan-kadkol) [reply packages/itwinui-css/src/input-container/input-container.scss:1]: I didn't think we need to change these magic numbers since the changes in this PR were applied to the container, and so the start button and icon just shift right by the same amount:

https://github.com/iTwin/iTwinUI/assets/45748283/865d9a19-1282-401a-bcd0-536312de1afc

I also double checked by taking screenshots:

|Before|With this PR|
|-|-|
|image|image|

Comment (mayank99) [reply packages/itwinui-css/src/input-container/input-container.scss:1]: I'm not sure I understand why it's not necessary to change these numbers. We removed the size-specific padding in `Icon`, so it makes sense to do the same here

Comment (rohan-kadkol) [reply packages/itwinui-css/src/input-container/input-container.scss:1]: Here is an illustration based on what I understood:

image

> We removed the size-specific padding in `Icon`, so it makes sense to do the same here

The size specific padding that we removed was from the `padded` prop in `Icon`. But the padding in the search icon is instead coming from `--_iui-flex-input-icon-padding` on `.iui-search-icon` (unchanged in this PR) ([code](https://github.com/iTwin/iTwinUI/pull/1828/files#diff-81de11dd7974a289c9463c407fbb06e31bb9e32575be8bb2e64cf3cbb1c8644dR47)).

Thus, I think these numbers don't need to be changed since we need to manually change the size/width of search-icon to match the size/width of the button; and the current numbers still achieve this.

Comment (mayank99) [inline packages/itwinui-css/src/input-container/input-container.scss:1]: I think we should also try to collapse the padding between an icon and an the `<input>` itself, because otherwise it adds up to a lot of empty space.

Current:
image

Expected:
image

(This will be slightly complicated, because the input could be used without an icon, but worth trying.)

cc @FlyersPh9

Comment (rohan-kadkol) [reply packages/itwinui-css/src/input-container/input-container.scss:1]: Sounds good, I think it might be better to work on that in a new PR.

If so, maybe we can release 3.5.0 without this PR. The next minor release could include this PR along with the PR to handle the above suggestion. That way we don't rush these two PRs and don't delay 3.5.0, and similar input container changes can come in the same release.

Comment (rohan-kadkol) [reply packages/itwinui-css/src/input-container/input-container.scss:1]: After our DMs, I realized what you were intending to highlight. Yes, we can make the .iui-search-icon use a constant padding instead of being dependent on the size. However, this will make the icon and the button no longer aligned:

image

As discussed, looks like we're okay with this. This is why I removed size-dependent padding on .iui-search-icon and instead used the `padded` on `Icon` (https://github.com/iTwin/iTwinUI/pull/1828/files/62f5e5747ea6e4977d341174f6357aa2a69bb18c..8e4104d94fa4d1c41e01825a39e1bb69638f7292).

Some issues arose because of this mismatch between the size/width of the icon and the IconButton such as the layout shift when toggling between the collapsed and the expanded states in an Expandable SearchBox:

https://github.com/iTwin/iTwinUI/assets/45748283/98cc8497-87f0-4430-9e93-b6d20c348270

I am currently working on this.

**Review by Ben-Pusey-Bentley — APPROVED** [review]
